### PR TITLE
added functions to allow for bonus manager

### PR DESCRIPTION
**Pull request by dharkmattr**

### CODE REVIEW

**Comment by tjcloa** [comment]
- No natspec provided - losing automatic docs generation option and makes it more difficult to review - no explicit intention  
- No unit tests tests - would facilitate review too

File: contracts/masset/BonusManager.sol [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        emit onMaxChanged(msg.sender, max);
+    }
+
+    function setMin(uint256 _min) public onlyOwner {
+        min = _min;
+        emit onMinChanged(msg.sender, min);
+    }
+
+    function setTokens(address[] memory _tokens) public onlyOwner {
+        for(uint i=0; i<tokens.length; i++) {
```
Comment (night-pm) [inline contracts/masset/BonusManager.sol:104]: Cache the length value of array. Also it's cost saving to use ++i.

**Review by night-pm — COMMENTED** [review]
Most of the comments are related to gas savings.

**Review by night-pm — APPROVED** [review]
LGTM

File: contracts/masset/BonusManager.sol [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    function setPaused(bool _paused) external onlyOwner {
+        paused = _paused;
+        emit onPausedChanged(msg.sender, paused);
+    }
+
+    /*****  Method to extract funds, owner only */
+
+    function extractFunds(address _tokenAddress, uint256 _amount) external onlyOwner {
+        IERC20 token = IERC20(_tokenAddress);        
+        require(token.transfer(owner(), _amount), "transfer failed");
```
Comment (tjcloa) [inline contracts/masset/BonusManager.sol:121]: recommendation: use OZ SafeERC20.safeTransfer to support tokens that return no value (and instead revert or throw on failure), non-reverting calls are assumed to be successful

File: contracts/masset/BonusManager.sol [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        // to prevent abuse, we make sure 
+        // that the total reward + bonus do not exceed to potential penalty
+        // so that users don't deposit and immediately withdraw to drain
+        // the bonus manager of funds
+
+        if(bonus + reward > penalty) {
+            bonus = penalty > reward ? penalty - reward : 0;
+        }
+
+        return bonus;
```
Comment (tjcloa) [inline contracts/masset/BonusManager.sol:169]: what if bonus > balance here?

**Comment by tjcloa** [comment]
1. CI is failing - check tests
2. see inline comments

**Review by tjcloa — CHANGES_REQUESTED** [review]
pls address comments

File: contracts/masset/Masset.sol [hunk 4]
```diff
@@ -523,11 +539,18 @@ contract Masset is IMasset, IERC777Recipient, InitializableOwnable, Initializabl
...
 
-    function migrateToV4() public {
-        require(keccak256(bytes(version)) == keccak256(bytes("3.0")), "wrong version");
+    // Temporary migration code
 
-        version = "4.0";
+    function migrateToV5() public {
+        require(keccak256(bytes(version)) == keccak256(bytes("4.0")), "wrong version");
+        version = "5.0";
+        bonusManager = IBonusManager(0);
```
[the remark this answers is not in the document]
Comment (dharkmattr) [reply contracts/masset/Masset.sol:552]: Yes, the bonus manager will cannot be set before the masset is upgraded, and the migrate function is part of the migration script.

File: contracts/masset/Masset.sol [hunk 5]
```diff
@@ -482,6 +494,10 @@ contract Masset is IMasset, IERC777Recipient, InitializableOwnable, Initializabl
         return address(rewardManager);
     }
 
+    function getBonusManager() external view returns (address) {
```
[the remark this answers is not in the document]
Comment (dharkmattr) [reply contracts/masset/Masset.sol:496]: Address is omitted to be consistent with the naming in this file, eg getBasketManager, getToken etc

**Review by tjcloa — APPROVED** [review]

**Pull request merged**